Project structural change to best practice and introduce unit tests

karthiks opened the pull request:
I see a opportunities for code improvements and have started to make changes to it. Please find the key changes as part of this merge request:
1. created top level `src` directory as peer to newly created `tests` directory. This structure should help with growing complexity of the project that are to be backed up with battery of tests - unit, integration, etc.
2. Started adding unit tests for `dhanhq.py` and in the course of it refactored some of its methods where common code can be extracted to private methods for better code readability and maintainability
3. Added `Dev_Readme.md` to aid faster dev setup to get going
4. Modularizing `dhanhq.py` has resulted in reduction of LoC by ~100 so far.

P.S: I do intend to add more unit tests and refactor the code accordingly. Expect more pull requests your way ;)

Let me know what you think.

karthiks commented:
@dhan-oss What is your take on this pull-request?

dhan-oss commented:
Hello @karthiks 

Loved what you did here, and we truly appreciate your contribution. The team will review this and merge the PR if everything works fine.

karthiks commented:
@dhan-oss Thanks for the prompt update! There is more to do and expect more PRs from me in this space.

karthiks commented:
@dhan-oss  FYI, `dhanhq.py` is now covered with unit tests except for a method or two, and is modularized with better scope for further refactoring. The Dev_README.md is updated with project structure visually. Cheers!

P.S: Where required, I have placed #ToDo comments or some marker comments on some bad code segments that needs fix, but those would be breaking changes, that can be done in next PR post discussions.

karthiks commented:
@dhan-oss Got a chance to run through the stuff?

karthiks commented:
@dhan-oss The earlier commit is code refactored on following counts as long description:

- [x] Extracted HTTP Methods to its own class for better modularity.
- [x] Extracted client-id and api-token to Context class signifying the client accessing the API. 
  This class would evolve to compose all connection mechanisms to API Server - HTTP and WebSocket.
- [x] Relevant Constants are grouped into Enum data-structure. 
  Note: There is more to do in this area, for other constants.

Hope, you are enjoying the code and design evolution

karthiks commented:
@dhan-oss Testing PRs shouldn't take this long, IMO. In the last commit, I have added integration tests to the project. The deal is that, if you have a stub server URL that can be used for integration tests, we can build integration tests upon the ones that I have just included to check for break in API contracts.

Let me know what is troubling you at your end.

Oh yeah, merry Christmas!!..

karthiks commented:
@dhan-oss FYI, Found and fixed a bug that was caught with addition of integration tests. The named param `data=` remained in order methods of `dhanhq` and that is removed with the simplified API (see specific commit with commit message on this above). Hope this helps the testing team's endeavors.

karthiks commented:
@dhan-oss The next set of kicker of refactoring on the code-base is getting cooked and is almost done. This is built on this and are you game for making DhanHQ SDK the hottest amongst its peers for its coding and UX standards?

dhan-oss commented:
Hello @karthiks 

We reviewed the PR and it looks good. Before merging to main branch, just wanted to let you know that we will not be setting this as the latest release for now and will do the same after a week. Also, would love if you can write the ReadMe too in the next PR, post that we will add new release on pypi as well.

Grateful to have you contribution and looking forward to more!

karthiks commented:
Sure, would be happy to write one. The customer facing `README.md` I believe needs to be updated with the **BREAKING CHANGE** information on introduction of `HTTPContext` to avoid repeated passing in of creds to different APIs of Dhan. Other than that much of the change is introduction of unit and integration tests to improve developer output in terms of speed and safety. The `DEV_README.md` is something I have already included, so the team and other contributors to the repo are on the same page on this engineering practice that today is supposedly a defacto.

karthiks commented:
@dhan-oss The `README.md` is now updated per your advice and the other APIs (`DhanFeed` and `OrderSocket` classes) in SDK as well leverage `DhanContext`. With this change, the release reaches a logical milestone, of leveraging `DhanContext` defined once by client and reused across their code, thus improving developer experience.

dhan-oss commented on README.md:
```diff
@@ -17,6 +17,40 @@ Not just this, you also get real-time market data via DhanHQ Live Market Feed.
+  | from dhanhq import marketfeed.DhanFeed | from dhanhq import DhanFeed |
+  | from dhanhq import orderupdate.OrderSocket | from dhanhq import OrderSocket |
+
+- The constants that were earlier part of modules `marketfeed` and `orderupdate` are now moved to its respective classes contained in the modules for better developer experience. 
```
On this, would prefer if the classes are called 'marketfeed' and 'orderupdate' itself, as it is more contextual then DhanFeed and OrderSocket.

karthiks replied:
@dhan-oss  I agree. In fact, I didn't change the package (i.e filename) and class name (i.e class in that file) in this PR because I was thinking of bringing in incremental changes than bringing in one sweeping change. In the next minor release, we can change these class names to make it more aligned to business. Even `dhanhq` class name is not pythonic that should be changed to more meaningful name like `DhanCore` or `DhanMain` or `DhanREST` or `DhanSync` etc.

Or should we rename classes in this release only? If so, which are the ones?

karthiks replied:
Just to be clear we should be changing our current 
"`dhanhq` directory" > "`dhanhq.py` file" > "`dhanhq` class" 
to something like 
"`dhanhq` directory" > "`dhan_core.py` file" > "`DhanCore` class", 
should we choose DhanCore as class name.

karthiks replied:
@dhan-oss Where are we on this? Did I make sense to you?

karthiks commented:
@dhan-oss When do I get to see this PR merged into main branch? And did you get a chance to review the next PR (see last commit there, which is built on top of the changes of this PR)?

dhan-oss commented:
Hey @karthiks 

Thank you for creating this PR and making much required structural changes. We would love to have your contributions moving forward to the library as well.

Best,
Team Dhan
